lib/maths/unittest/CDataFrameUtilsTest.cc

```diff
@@ -534,51 +534,66 @@ BOOST_AUTO_TEST_CASE(testStratifiedCrossValidationRowMasks) {
+        std::tie(std::ignore, testingRowMasks, std::ignore) =
+            maths::CDataFrameUtils::stratifiedCrossValidationRowMasks(
+                1, *frame, 0, rng, numberFolds[0], numberBins, allTrainingRowsMask);
 
-        BOOST_REQUIRE_EQUAL(numberFolds[0], trainingRowMasks.size());
-        BOOST_REQUIRE_EQUAL(numberFolds[0], testingRowMasks.size());
+        TDoubleVecVec targetDecile(numberFolds[0], TDoubleVec(numberBins));
 
         core::CPackedBitVector allTestingRowsMask{numberRows, false};
         for (std::size_t fold = 0; fold < numberFolds[0]; ++fold) {
-            // Expect count ~ binomial(n, 1 / "number folds").
+            // Count should be very nearly the expected value.
             double expectedTestRowCount{allTrainingRowsMask.manhattan() /
                                         static_cast<double>(numberFolds[0])};
             BOOST_REQUIRE_CLOSE_ABSOLUTE(expectedTestRowCount,
-                                         testingRowMasks[fold].manhattan(),
-                                         3.0 * std::sqrt(expectedTestRowCount));
+                                         testingRowMasks[fold].manhattan(), 10.0);
             BOOST_REQUIRE_EQUAL(0.0, testingRowMasks[fold].inner(allTestingRowsMask));
-            BOOST_REQUIRE_EQUAL(0.0, trainingRowMasks[fold].inner(testingRowMasks[fold]));
-            BOOST_REQUIRE_EQUAL(trainingRowMasks[fold].manhattan(),
-                                trainingRowMasks[fold].inner(allTrainingRowsMask));
             BOOST_REQUIRE_EQUAL(testingRowMasks[fold].manhattan(),
                                 testingRowMasks[fold].inner(allTrainingRowsMask));
             allTestingRowsMask |= testingRowMasks[fold];
+
+            TDoubleVec values;
+            frame->readRows(1, 0, frame->numberRows(),
+                            [&](core::CDataFrame::TRowItr beginRows,
+                                core::CDataFrame::TRowItr endRows) {
+                                for (auto row = beginRows; row != endRows; ++row) {
+                                    values.push_back((*row)[0]);
+                                }
+                            },
+                            &testingRowMasks[fold]);
+            std::sort(values.begin(), values.end());
+            for (std::size_t i = 1; i < numberBins; ++i) {
+                targetDecile[fold][i] = values[(i * values.size()) / numberBins];
+            }
+        }
+
+        for (std::size_t i = 1; i < numberBins; ++i) {
+            TMeanVarAccumulator testTargetDecileMoments;
+            for (std::size_t fold = 0; fold < numberFolds[0]; ++fold) {
+                testTargetDecileMoments.add(targetDecile[fold][i]);
+            }
+            LOG_DEBUG(<< "variance in test set target percentile = "
+                      << maths::CBasicStatistics::variance(testTargetDecileMoments));
+            BOOST_REQUIRE(maths::CBasicStatistics::variance(testTargetDecileMoments) < 0.02);
```

Comment:
It's worth noting that `BOOST_TEST_REQUIRE` here would print the values either side of the `<` in the event of failure, whereas `BOOST_REQUIRE` won't.

It seems that `BOOST_REQUIRE(condition)` is the equivalent of `BOOST_TEST_REQUIRE((condition))`.  I wasn't aware of this until now, and `BOOST_REQUIRE` is the cleaner solution for complex conditions where diagnostics cannot be printed.  But for this line the improved diagnostics _could_ be printed by `BOOST_TEST_REQUIRE`.

Reply:
I changed this in #788

Reply:
Thanks